### ISSUE TITLE
Design OOV/Rare Root Handling Strategy

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#46]
## Problem

After analyzing the M1 corpus (Issue #41), we found:
- **Total unique roots**: 2,098,172
- **Top 5K roots**: 81.2% coverage
- **Remaining 2M+ roots**: 18.8% of corpus (proper nouns, numbers, rare words)

Our morpheme-aware embeddings (Issue #43) use a 5K root vocabulary. We need a strategy for handling the 18.8% of words that use rare/OOV roots.

### Option 1: Character-Level Fallback
- For OOV roots, use character-level encoding
- Compose: char_embed(root) + semantic_affixes + grammatical_features
- **Pro**: Handles any root, no <UNK>
- **Con**: Character semantics may be weak

### Option 2: Subword Tokenization (BPE/WordPiece)
- Break rare roots into subword units
- Learn subword embeddings for common patterns
- **Pro**: Better than pure characters
- **Con**: Less interpretable, more complex

### Option 3: Proper Noun Detection + Special Handling
- Detect proper nouns (capitalized, not in dictionary)
- Use entity-type embeddings (person, place, organization)
- **Pro**: Semantically meaningful for names
- **Con**: Requires entity type classification

### Option 4: <UNK> with Context
- Map all OOV roots to <UNK>
- But compose with known affixes + grammatical features
- **Pro**: Simple, preserves morphological structure
- **Con**: Loses root semantics entirely

### Option 5: Expand Vocabulary to 10K
- Use top 10K roots instead of 5K
- Coverage increases to 85.3%
- **Pro**: Simple, better coverage
- **Con**: More parameters (1.28M vs 640K for roots)

## Evaluation Criteria

Test on sentences containing rare roots:
1. **Compositional understanding**: Does model understand affixes even with OOV root?
2. **Context appropriateness**: Can model infer rough semantics from context?
3. **Task performance**: Impact on retrieval/QA accuracy

## Priority

**Medium** - Can defer until after initial training (Issue #45)
- 81.2% coverage is good for M1
- Can add rare root handling in later iteration

## Next Steps

1. Collect statistics on rare root types (proper nouns vs rare words)
2. Implement 2-3 options
3. Evaluate on test set
4. Choose best approach for M1

## Related

- Issue #41: Root frequency analysis
- Issue #43: MorphemeAwareEmbedding architecture
- Issue #45: Training on M1 corpus

**Comment by marctjones** [comment]
## ✅ Recommended Strategy: Proper Noun Detection + Hash Fallback

After analyzing OOV handling trade-offs and their impact on vocabulary size decisions, here's the recommended approach for M1:

### Architecture

```python
class MorphemeAwareEmbedding(nn.Module):
    def __init__(self, root_vocab_size=5000, num_entity_types=10, num_hash_buckets=500):
        # Standard root embeddings (5K × 128d = 640K params)
        self.root_embed = nn.Embedding(root_vocab_size, 128)
        
        # OOV handling (65K params total)
        self.entity_embed = nn.Embedding(num_entity_types, 128)  # 10 × 128 = 1.3K
        self.hash_embed = nn.Embedding(num_hash_buckets, 128)    # 500 × 128 = 64K
    
    def get_root_embedding(self, root: str) -> torch.Tensor:
        # 1. In-vocabulary: use learned embedding
        if root in self.root_vocab:
            return self.root_embed(self.root_vocab[root])
        
        # 2. Proper noun: use entity type embedding
        if self.is_proper_noun(root):
            entity_type = self.classify_entity(root)
            return self.entity_embed(entity_type)
        
        # 3. Other OOV: hash to bucket
        hash_idx = hash(root) % self.num_hash_buckets
        return self.hash_embed(hash_idx)
```

### Why This Strategy?

**Parameter Budget**:
- Total: ~775K params (640K roots + 69K affixes/composer + 65K OOV)
- Comparable to 10K vocab alone (1.28M), but better OOV handling
- Only +65K params over baseline 5K vocab

**Coverage**:
- 81.2% in-vocabulary (5K roots)
- Excellent OOV recovery for 18.8%:
  - Proper nouns (majority): entity-type embeddings
  - Technical terms/rare words: hash buckets
  - Better than single `<UNK>` token

**Performance**:
- Fast inference: O(1) hash lookup
- Simple detection: capitalization + heuristics
- Handles Wikipedia corpus well (many proper nouns)

**Implementation Complexity**:
- ~100 lines of code
- Simple heuristics for entity classification
- No external dependencies

### Entity Types (10 categories)

1. PERSON - "Einstein", "Zamenhof"
2. PLACE - "Parizo", "Tokio"
3. ORGANIZATION - "UN", "UNESCO"
4. LANGUAGE - "angla", "Esperanto"
5. NATIONALITY - "Amerika", "Ĉina"
6. EVENT - "Olimpiko", "Mondmilito"
7. PRODUCT - "Android", "Windows"
8. ARTWORK - "Hamlet", "Mona Lisa"
9. SCIENTIFIC - "DNA", "H2O"
10. OTHER - unclassified proper nouns

### Hash Buckets (500 buckets)

- Technical terms: "biologi", "kemio"
- Rare words: archaic, specialized vocabulary
- Compound roots: multi-part constructions
- Hash distributes OOV roots across buckets
- Provides better fallback than `<UNK>`

### Comparison to Alternatives

| Strategy | Total Params | OOV Quality | Complexity | Best For |
|----------|--------------|-------------|------------|----------|
| **Proper Noun + Hash** ⭐ | **775K** | **Very Good** | **Low** | **M1 Wikipedia** |
| 10K vocab + `<UNK>` | 1,349K | Weak | Very Low | Simple architecture |
| 5K vocab + Char-level | 759K | Good | Medium | Morphologically rich roots |
| 10K vocab + Proper Noun | 1,351K | Good | Low | Higher in-vocab coverage |
| 5K vocab + Context | ~800K | Excellent | High | M2/M3 with reasoning core |

### Implementation Plan

**Phase 1: Extend MorphemeAwareEmbedding** (2 hours)
- Add entity and hash embeddings
- Implement proper noun detection
- Implement entity classification heuristics

**Phase 2: Training Integration** (1 hour)
- OOV embeddings trained alongside root embeddings
- Use masked root prediction for OOV buckets
- Entity types learn from context

**Phase 3: Evaluation** (1 hour)
- Test on sentences with proper nouns
- Test on technical terms
- Compare to `<UNK>` baseline

**Total**: 4 hours implementation

### Next Steps

1. Update Issue #43 MorphemeAwareEmbedding with OOV handling
2. Implement proper noun detection
3. Implement entity classification
4. Add hash fallback
5. Test on M1 corpus samples
6. Integrate into training script (Issue #44)

This strategy provides the best balance of **parameter efficiency**, **OOV quality**, and **implementation simplicity** for M1.

See [[Compositional-Embeddings#OOV Handling Strategy]] wiki for full details.

**Comment by marctjones** [comment]
## Related Finding: Missing Suffix Training Due to Root Vocab Gaps

During affix training analysis, discovered that several suffixes have corpus occurrences but no training samples because their roots aren't in the root vocabulary:

| Suffix | Corpus Occurrences | Example Missing Roots |
|--------|-------------------|----------------------|
| -ism   | 19,008 | esper, social, naci, komunik |
| -estr  | 6,676 | ŝip, regn, detal |
| -ing   | 5,159 | cigar, plum, fingr |
| -uj    | 4,276 | patr, pom, monat |

This is a concrete example of the OOV problem: **common affixed words can't be embedded compositionally if their roots are missing**.

### Proposed Solution
When building root vocabulary, ensure high-frequency affixed words contribute their roots even if the bare root is rare in corpus.

Example: "esperanto" is common, so "esper" should be in root vocab even if "espera" is rare.
